### ISSUE TITLE
generate() produces garbage despite low train loss: prefill skips causal masking when a KV cache is attached

### ISSUE THREAD

**Issue by Mikyx-1** [Mikyx-1/TinyLLM#17]
## Summary

`model/attention.py`'s `MultiHeadAttention.forward` only applies the causal mask `if cache is None`. That's correct for a single-token decode step (nothing to mask — one new token attending to its cached past is trivially causal), but `generate()`'s **prefill** call passes the *entire multi-token prompt* through with a cache already attached (`model/generate.py:37`, `caches` is non-`None` even though `T_new == prompt_len > 1`). So prefill skips masking entirely, and every prompt token attends to *future* prompt tokens as well as past ones.

That non-causal contamination gets baked into the K/V tensors written into the cache, since the corrupted hidden state at each position propagates into what gets cached for later layers. Every decode step afterward attends against that corrupted cache, so generation diverges — often after just one or two tokens — even when the underlying model is correct.

## Repro

Trained a checkpoint to train loss ≈0.06 on `data/sft_dataset.json` (heavy SFT/memorization run). Teacher-forced forward passes (the same code path used to compute training loss — never uses a cache) correctly predict every token of a memorized answer. But `generate()`/`inference.py` on the *same checkpoint* with the *same prompt* produces unrelated, garbled text after the first generated token.

Direct isolation:
\`\`\`python
# Path A: full forward, no cache, over prompt+first_generated_token -> ground truth for step 2
logits_full2, _ = model(full_seq, caches=None, start_pos=0)
pred_full2 = logits_full2[0, -1].argmax().item()   # -> "name" (correct, matches training data)

# Path B: KV-cache decode step, using the cache populated by prefill -- same weights, same context
logits_dec, _ = model(next_tok, caches=caches, start_pos=len(prompt_ids))
pred_dec = logits_dec[0, -1].argmax().item()        # -> "wife" (wrong)
\`\`\`
Same weights, same effective context — different answer. The only difference is that Path B's cache was populated by a non-causally-masked prefill.

## Root cause

\`\`\`python
if cache is None:
    scores = scores.masked_fill(self.causal_mask[:T_new, :T_new], float("-inf"))
\`\`\`
This branches on "is a cache present" when it should branch on "are there multiple new query positions that could see each other's future." Those are different conditions: prefill has `cache is not None` **and** `T_new > 1`.

## Why training was unaffected

`model(x, targets=y)` during training never passes a cache, so it always hit the `cache is None` branch — loss numbers were always legitimate. This bug only ever affected `generate()`/inference.

## Fix

Mask only the new-token block against itself; cached past positions are unconditionally causal-valid (they precede every new token by construction) and never need masking:
\`\`\`python
if T_new > 1:
    causal = self.causal_mask[:T_new, :T_new]
    scores[..., :, T_past:] = scores[..., :, T_past:].masked_fill(causal, float("-inf"))
\`\`\`
where `T_past` is the cache length *before* this call's tokens were appended. This is a no-op change for training (`T_past == 0`, masks the whole `T_new × T_new` block, identical to before) and for single-token decode (`T_new == 1`, skipped, as before) — it only changes behavior for multi-token prefill, which is exactly the previously-broken case.

Verified after the fix: the isolation test above now returns "name" for both Path A and Path B, and `inference.py` reproduces memorized answers exactly instead of diverging after the first token.
